### ISSUE TITLE
Training stuck on first epoch

### ISSUE THREAD

**Issue by alexisdrakopoulos** [mlfoundations/open_clip#1016]
Hi, I tried training using:

```
python -m open_clip_train.main \
    --save-frequency 1 \
    --zeroshot-frequency 1 \
    --dataset-type webdataset \
    --train-data 'tar_files/dataset_{000..042}.tar'  \
    --train-num-samples 788603 \
    --warmup 10000 \
    --batch-size=256 \
    --lr=1e-3 \
    --wd=0.1 \
    --epochs=200 \
    --workers=1 \
    --model ViT-B-16-SigLIP
```

and I get the nice output:
```
...
2025-01-02,19:49:58 | INFO | Created AdamW (adamw) optimizer: lr: 0.001, betas: (0.9, 0.98), eps: 1e-06, weight_decay: 0.1, amsgrad: False, foreach: None, maximize: False, capturable: False, differentiable: False, fused: None
loading file spiece.model from cache at None
loading file tokenizer.json from cache at /root/.cache/huggingface/hub/models--timm--ViT-B-16-SigLIP/snapshots/41f575766f40e752fdd1383e9565b7f02388c1c4/tokenizer.json
loading file added_tokens.json from cache at None
loading file special_tokens_map.json from cache at /root/.cache/huggingface/hub/models--timm--ViT-B-16-SigLIP/snapshots/41f575766f40e752fdd1383e9565b7f02388c1c4/special_tokens_map.json
loading file tokenizer_config.json from cache at /root/.cache/huggingface/hub/models--timm--ViT-B-16-SigLIP/snapshots/41f575766f40e752fdd1383e9565b7f02388c1c4/tokenizer_config.json
loading file chat_template.jinja from cache at None
2025-01-02,19:49:58 | INFO | Start epoch 0
Epoch 0:   0%|                                                                                                                    | 0/3081 [00:00<?, ?it/s]  
```

but it never completes a single batch nor continues. It's been 20 minutes. Furthermore running nvidia-smi reveals the GPU isn't being used at all:

```
Thu Jan  2 20:07:33 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.78                 Driver Version: 550.78         CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4090        On  |   00000000:82:00.0 Off |                  Off |
|  0%   31C    P8             17W /  450W |    1247MiB /  24564MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
+-----------------------------------------------------------------------------------------+
```

even though device cuda is selected. Any ideas? Could it be a problem with my webdataset?

The hardware setup is:

- AMD EPYC 7713 64-Core Processor
- 512GB DDR4
- RTX 4090
- Samsung SSD 990 PRO 2TB

**Comment by rwightman** [comment]
@alexisdrakopoulos most likely the dataset... the dataset code isn't particularly modular here so can be a bit of hacking to isolate bits and walk through. Make sure you're not on the current webdatset version, 0.2.86 is the most recent tested version that does not introduce regressions.

I worked on another project last year that I've shelved for now, 'chug' (`pip install git+https://github.com/huggingface/chug.git`) that has dataloading components that will read openclip compatible image-text webdatasets. It's modular so it might be a bit easier to use that to check the tars are as expected than to finagle args and whatnot in open_clip's data routines.

```
import chug
img_cfg = chug.ImageInputCfg(size=(512, 512), transform_type='image_timm')
img_fn = chug.create_image_preprocessor(input_cfg=img_cfg, is_training=True)
txt_fn = lambda x: x  # no tokenizer
task_cfg = chug.DataTaskImageTextCfg(
    image_process_fn=img_fn,
    text_process_fn=txt_fn,
)
data_cfg = chug.DataCfg(
    source='/data/cc12m/cc12m-train-{0000..2175}.tar',
    batch_size=8,
    num_samples=10968539,
    format='wds',
)
lb = chug.create_loader(
    data_cfg,
    task_cfg,
    is_training=True,
)
ii = iter(lb)
sample = next(ii)
```

sample would be a dict there and there should be an 'image' and 'text' key if it's successful.

**Comment by rwightman** [comment]
Also, if you have really large images, you can end up in a situation where you have CPUs spinning their wheels at 100% taking forever to decode & preprocess images, thus slowing training down incredibly..

**Comment by alexisdrakopoulos** [comment]
My images are 512x512. I should test out the data loader separately. I did modify it slightly:

```python
tokenizer_apply = lambda text: tokenizer(text["description"])[0]
pipeline.extend(
        [
            wds.select(filter_no_caption_or_no_image),
            wds.decode("pilrgb", handler=log_and_continue),
            wds.rename(image="jpg;png;jpeg;webp", text="json"),
            wds.map_dict(image=preprocess_img, text=tokenizer_apply),
            wds.to_tuple("image", "text"),
            wds.batched(args.batch_size, partial=not is_train),
        ]
    )
```

because my webdataset returns JSON. That shouldn't be the issue though...

**Comment by alexisdrakopoulos** [comment]
Ah actually it might be filter_no_caption_or_no_image. It is hardcoded to look for a "txt" entry. I need to look at the pipeline more carefully. Strange that it just hangs instead of erroring.

**Comment by rwightman** [comment]
@alexisdrakopoulos webdataset has a default decoder for many keys, including 'json', unless you disable it, it will deserialize the json into an obj automatically. You might be conflicting with that if you try to rename the json to text.

And yeah the filtering will look for the text key

**Comment by rwightman** [comment]
filtering was added to avoiding crashing very expensive training runs where some samples had missing captions... but if there are NO captions it will try and iterate through the whole dataset before it barfs... maybe end up getting stuck though as that probably hasn't been tested :)

**Comment by rwightman** [comment]
print debugging is a common goto here, inject wds.map into various pipeline points, map to a lambda or fn where you print sample keys, contents, add breakpoints, etc

**Comment by alexisdrakopoulos** [comment]
@rwightman  I have one more question, I was trying to follow https://github.com/mlfoundations/open_clip/discussions/812 

It mostly makes sense except for the part where it asks for the model path. How can I find the model path when loading for example a timm sigLIP model? It downloads it to some cache but it's very opaque, I can't figure out where the .pt file is to pass as a pre-trained starter

**Comment by rwightman** [comment]
@alexisdrakopoulos it's easiest to rely on the pretrain tags and let the hf hub / cache handle the download.

So `--pretrained webli` with that B/16 model would give you this checkpoint https://huggingface.co/timm/ViT-B-16-SigLIP

https://github.com/mlfoundations/open_clip/blob/b2f1403605aade5a004434076246b6bc741aa47d/src/open_clip/pretrained.py#L475-L477

If you're in an offline env, still easier to use the HF hub mechanism rather than downloading tokenizer and pth files manually, but you can use the `cache_dir` arg to override the root folder where those files get downloaded, then you can move them specifically to another machine / container, etc and use cache_dir again ... make sure HF_HUB_OFFLINE=1 set to avoid a check for updates.

**Comment by rwightman** [comment]
If you need specific control over loading image and text checkpoints separately, it'll require some hacks but others have done it and there were some old PR that got stale.

**Comment by alexisdrakopoulos** [comment]
Thanks! I'll try a command somewhat like:

python -m open_clip_train.main \
    --save-frequency 1 \
    --zeroshot-frequency 1 \
    --dataset-type webdataset \
    --train-data 'tar_files/dataset_{000..042}.tar'  \
    --train-num-samples 788603 \
    --warmup 10000 \
    --batch-size=256 \
    --lr=1e-3 \
    --wd=0.1 \
    --epochs=200 \
    --workers=1 \
    --model ViT-B-16-SigLIP \
    --pretrained webli

after I fix the text issue. You've been really helpful thank you.

**Comment by alexisdrakopoulos** [comment]
Also happy new year!
